### PR TITLE
chore: Fix tests

### PR DESCRIPTION
**Pull request by eps1lon**
Fixing 

```
[test] FAIL src/__tests__/render.js (9.164 s)
[test]   ● render API › legacyRoot uses legacy ReactDOM.render
[test] 
[test]     TypeError: jestDiff is not a function
[test] 
[test]       161 |           errorMessage =
[test]       162 |             'Unexpected warning recorded: ' +
[test]     > 163 |             jestDiff(expectedMessages[0], normalizedMessage)
[test]           |             ^
[test]       164 |         } else {
[test]       165 |           errorMessage =
[test]       166 |             'Unexpected warning recorded: ' +
[test] 
[test]       at console.jestDiff (tests/toWarnDev.js:163:13)
[test]       at printWarning (node_modules/react-dom/cjs/react-dom.development.js:105:30)
[test]       at error (node_modules/react-dom/cjs/react-dom.development.js:79:7)
[test]       at Object.render (node_modules/react-dom/cjs/react-dom.development.js:37824:5)
[test]       at Object.render (src/pure.js:133:16)
[test]       at render (src/pure.js:152:12)
[test]       at callback (src/act-compat.js:42:24)
[test]       at act (node_modules/react/cjs/react.development.js:2719:16)
[test]       at actImplementation (src/act-compat.js:[41](https://github.com/testing-library/react-testing-library/actions/runs/8349938959/job/22855248000?pr=1287#step:7:42):25)
[test]       at renderRoot (src/pure.js:1[45](https://github.com/testing-library/react-testing-library/actions/runs/8349938959/job/22855248000?pr=1287#step:7:46):6)
[test]       at renderRoot (src/pure.js:241:10)
[test]       at src/__tests__/render.js:218:13
[test]       at Object.callback (tests/toWarnDev.js:185:9)
[test]       at __EXTERNAL_MATCHER_TRAP__ (node_modules/expect/build/index.js:325:30)
[test]       at Object.throwingMatcher [as toErrorDev] (node_modules/expect/build/index.js:326:15)
[test]       at Object.toErrorDev (src/__tests__/render.js:219:8)
```

-- https://github.com/testing-library/react-testing-library/actions/runs/8349938959/job/22855248000?pr=1287#step:7:43

Also adjusts to changed warnings in https://github.com/facebook/react/pull/28477

### CODE REVIEW

File: tests/toWarnDev.js [hunk 1]
```diff
@@ -29,9 +29,12 @@ SOFTWARE.
 /* eslint-disable func-names */
 /* eslint-disable complexity */
 const util = require('util')
-const jestDiff = require('jest-diff').default
+const jestDiff = require('jest-diff').diff
```
Comment (eps1lon) [inline tests/toWarnDev.js:32]: No idea why this ever worked. It didn't change in `jest-diff` between 29.4.1 and 29.7.0 so it must have been some other thing that made it work incidentally.

**Comment by eps1lon** [comment]
Will fix codecov in a follow-up. Manual inspection shows 100%: https://app.codecov.io/github/testing-library/react-testing-library/commit/c4b23e4fd087fd27b95b2051de5fe156f7a07523

File: src/__tests__/render.js [hunk 2]
```diff
@@ -3,6 +3,13 @@ import ReactDOM from 'react-dom'
 import ReactDOMServer from 'react-dom/server'
 import {fireEvent, render, screen, configure} from '../'
 
+// Needs to be changed to 19.0.0 once alpha started.
+const isReactExperimental = React.version.startsWith('18.3.0-experimental')
+const isReactCanary = React.version.startsWith('18.3.0')
```
Comment (MatanBobi) [inline src/__tests__/render.js:8]: Shouldn't we check for `React.version.startsWith('18.3.0-canary')` here?

File: src/__tests__/render.js [hunk 3]
```diff
@@ -213,27 +220,35 @@ describe('render API', () => {
...
       render(<div />, {legacyRoot: true})
     }).toErrorDev(
-      [
-        "Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot",
-      ],
+      isReactCanary
+        ? [
+            "Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://react.dev/link/switch-to-createroot",
+          ]
+        : [
```
[the remark this answers is not in the document]
Comment (eps1lon) [reply src/__tests__/render.js:231]: That's what React is logging

File: src/__tests__/render.js (hunk 2 above)
Comment (eps1lon) [reply src/__tests__/render.js:8]: yep

**Pull request merged**